File: PRACTICA/Java/irrgarten/LabyrinthCharacter.java
package irrgarten;

public abstract class LabyrinthCharacter {
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;

    public LabyrinthCharacter(String name, float intelligence, float strength, float health){
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = health;
        row = 0;
        col = 0;
    }

    public LabyrinthCharacter(LabyrinthCharacter other){
        name = other.name;
        intelligence = other.intelligence;
        strength = other.strength;
        health = other.health;
        row = other.row;
        col = other.col;
    }

    public boolean dead(){
        return health < 0;
    }

    public int getRow(){
        return row;
    }

    public int getCol(){
        return col;
    }

    protected float getIntelligence(){
        return intelligence;
    }

    protected float getStrength(){
        return strength;
    }

    protected float getHealth(){
        return health;
    }

    protected void setHealth(float health){
        this.health = health;
    }

    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }

    @Override
    public String toString(){
        return name + " (Int: " + intelligence + ", Str: " + strength + ", Hp: " + health + ")";
    }

    protected void gotWounded(){
        health--;
    }

    public abstract float attack();
    public abstract boolean defend (float attack);
}

File: PRACTICA/Ruby/Irrgarten/labyrinth_character.rb
class Labyrinth_character
  def initialize(name_or_other, intelligence = nil, strength = nil, health = nil)
    if name_or_other.is_a?(Labyrinth_character)
      other = name_or_other
      @name = other.name
      @intelligence = other.intelligence
      @strength = other.strength
      @health = other.health
      @row = other.row
      @col = other.col
    else
      @name = other
      @intelligence = intelligence
      @strength = strength
      @health = health
      @row = 0
      @col = 0
    end
  end

  def dead
    return @health < 0
  end

  def set_pos(row, col)
    @row = row
    @col = col
  end

  def to_s
    "#{@name} (Int: #{@intelligence}, Str: #{@strength}, Hp: #{@health})"
  end

  protected

  def intelligence
    return @intelligence
  end

  def strength
    return @strength
  end

  def health
    return @health
  end

  def health=(value)
    @health = value
  end

  def got_wounded
    @health -= 1
  end

  public

  def attack
  end

  def defend(received_attack)
  end

  private_class_method :new
end